Prevent stalls in query_partition_key_range_concurrent

bhalevy opened the pull request:
Prevent stalls caused by query_partition_key_range_concurrent
nested calls when it never yields.

Fixes #14008

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2752/

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5527,10 +5527,10 @@ bool storage_proxy::is_worth_merging_for_range_query(
 storage_proxy::query_partition_key_range_concurrent(storage_proxy::clock_type::time_point timeout,
         locator::effective_replication_map_ptr erm,
-        std::vector<foreign_ptr<lw_shared_ptr<query::result>>>&& results,
+        std::vector<foreign_ptr<lw_shared_ptr<query::result>>>& results,
```
It's not really needed. The caller holds the variable ready.

It's clearer to convert this to a value parameter. To track the lifetime, you don't need to go up to the caller, instead you see it's safe here.

btw after all this we can probably merge the caller and callee.

avikivity commented:
The change is unsatisfying - it leaves the function a mix of coroutines and continuations. Further, it converts tail recursion to non-tail-recursion, so now there's a long trail of coroutines in memory awaiting the last call. The recursion should be converted to iteration.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2753/

bhalevy commented:
In v2 ([6614ae2](https://github.com/scylladb/scylladb/commit/6614ae2f30a6f0563d54fc6e9895b3c09739a965)):
- turn tail-recursion into iteration

I didn't fold `query_partition_key_range_concurrent` into `query_partition_key_range` since:
a. It's out of scope for this fix
b. It makes the code complex and I'm not sure that saving the extra coroutine is worth it.

I'll see if there's and visible performance benefit to that and submit a new PR if there is one.

avikivity commented:
> In v2 ([6614ae2](https://github.com/scylladb/scylladb/commit/6614ae2f30a6f0563d54fc6e9895b3c09739a965)):
> 
> * turn tail-recursion into iteration
> 
> I didn't fold `query_partition_key_range_concurrent` into `query_partition_key_range` since: a. It's out of scope for this fix b. It makes the code complex and I'm not sure that saving the extra coroutine is worth it.


The idea behind the suggestion was to simplify. If it doesn't work, then it doesn't work.

> 
> I'll see if there's and visible performance benefit to that and submit a new PR if there is one.


It will save an allocation, but it will be invisible. Scans have tons of allocations.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5694,27 +5695,19 @@ storage_proxy::query_partition_key_range_concurrent(storage_proxy::clock_type::t
+        auto error = bo::failure(std::move(wrapped_result).assume_error()).error();
+        auto ex = error.as_exception_future().get_exception();
+        p->handle_read_error(std::move(error), true);
+        co_return coroutine::return_exception_ptr(std::move(ex));
```
I'm sure this could be improved. Since this function returns a `result<>`, it could be propagated.

But, let's not mix the improvement with this series.

bhalevy replied:
I'd do that if it wasn't for the call to `handle_read_error`...

bhalevy replied:
Hmm, this isn't so bad:
```c++
            auto error = std::move(wrapped_result).assume_error();
            p->handle_read_error(error.clone(), true);
            co_return error;
```

testing...

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5694,27 +5695,19 @@ storage_proxy::query_partition_key_range_concurrent(storage_proxy::clock_type::t
+    }
+
+        // FIXME: indentation
+        foreign_ptr<lw_shared_ptr<query::result>> result = std::move(wrapped_result).value();
```
You could have used assume_value() since you know it's not an error. But it doesn't matter here.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5690,6 +5690,8 @@ storage_proxy::query_partition_key_range_concurrent(storage_proxy::clock_type::t
             exec.push_back(::make_shared<never_speculating_read_executor>(schema, cf.shared_from_this(), p, erm, cmd, std::move(range), cl, std::move(filtered_endpoints), trace_state, permit, std::monostate()));
             ranges_per_exec.emplace(exec.back().get(), std::move(merged_ranges));
+
+            co_await coroutine::maybe_yield();
```
I don't know if this is enough to fix the stall.  Note there is another loop on `i` above, so if we spend most of our iteration time there, we won't hit the maybe_yield here.

With 3 nodes it's likely everything was contiguous with RF=3.

bhalevy replied:
good point, I'll add one in the inner loop (that is execute only if the replication strategy does not `uses_tablets`).

bhalevy replied:
fixed in v3

avikivity commented on service/storage_proxy.cc:
```diff
@@ -14,6 +14,8 @@
 #include "partition_range_compat.hh"
```
nit: missing ':' in subject of first patch

avikivity commented:
Looks much better, and it makes visible that the patch may not be enough.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2757/

bhalevy commented:
In v3 ([77c354b](https://github.com/scylladb/scylladb/commit/77c354b7e585fb831021dfec670d8a63a56d9b11))
- rebased
- cleaned up patch subject
- cleaned up stray #include directives that were auto-added by the editor.
- change maybe_yield structure to have one in the outer loop and one in the inner-loop (and spaced apart so they's make sense if both are executed)

bhalevy commented:
In v4 ([e71e5d3](https://github.com/scylladb/scylladb/commit/e71e5d3fdab44b4b2867d18d21c08ccb4834c8e2)):
- query_partition_key_range_concurrent: do not return exception on error path, but rather copy the error `result` as per https://github.com/scylladb/scylladb/pull/14884#discussion_r1278615082.

denesb commented on service/storage_proxy.cc:
```diff
@@ -5573,6 +5573,8 @@ storage_proxy::query_partition_key_range_concurrent(storage_proxy::clock_type::t
             std::vector<dht::token_range> merged_ranges{to_token_range(range)};
```
[commitlog]: The calls are not nested anymore.

bhalevy commented:
updated changelog message

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2767/
